### ISSUE TITLE
ACME sanity check

### ISSUE THREAD

**Issue by arp242** [boringproxy/boringproxy#30]
Right not a failing ACME is kind of confusing:

> 2021/01/05 14:13:04 admin.mick-ci.com: obtaining certificate: [admin.mick-ci.com] Obtain: [admin.mick-ci.com] solving challenges: admin.mick-ci.com: no solvers available for remaining challenges (configured=[tls-alpn-01] offered=[http-01 dns-01 tls-alpn-01] remaining=[http-01 dns-01]) (order=https://acme-v02.api.letsencrypt.org/acme/order/108447721/7139455689) (ca=https://acme-v02.api.letsencrypt.org/directory)

The reason for this is that `admin.mick-ci.com` exists, but doesn't end up on the boringproxy server since I haven't modified the Varnish proxy yet.

It might be better to check for this first, before it even attempts to set up ACME. Something like `$(curl http://admin.domain:7000/test)" == "boringproxy"`. This was a clearer message can be presented, and it avoids hitting in to ACME rate limits.

**Comment by anderspitman** [comment]
There's actually still some higher level design work that needs to be done here. One problem with the current system is that any user on a boringproxy server instance can create a tunnel for any domain pointed at that instance. This is a security vulnerability. For example, imagine if I point *.example.com at the demo instance. Another user could create tunnels for any subdomain on example.com. Even worse, if I point the apex example.com at the instance, another user may be able to hijack it before I set up my tunnel.

I've considered a couple solutions to this:

1. Require users to add special DNS TXT records to tie domains to their specific boringproxy account. In this case boringproxy would have to do a DNS check to verify a proper TXT record before creating the tunnel.
2. Require users to add domains from the boringproxy interface before using them.

I've been leaning towards option 2, because it enables you to integrate boringproxy directly with DNS provider APIs. So not only can it set up tunnels, it can handle DNS for you so you don't have to log in to your DNS. You could even integrate domain purchasing. This is part of the original vision for boringproxy. However in practice (for me at least) I rarely need to change DNS settings, so I'm not sure it's worth the added complexity to boringproxy. To really be secure with option 2, you'd probably still need some level of DNS integration (either TXT records or API key) to ensure that the boringproxy user has control over the domain.

@arp242 thoughts?

**Comment by arp242** [comment]
But people would need access to the server in order to do that, right?

I'm not sure if this is really a security issue in boringproxy as such, but rather in the ACME protocol: it assumes that if you have access to whatever a domain points at, you have rightful control over that domain. I think that's mostly a reasonable assumption. Personally I wouldn't worry too much about this, if I understand the issue correct.

The big downside of the dns-01 verification is that you need the modify the DNS records; to do this automatically you need access to the DNS host's API, and very few have good scoping (i.e. the API keys allows modifying everything, you can't scope the to just one record). For the *.goatcounter.com certificate (wildcards require dns-01 verification) I ... just update it manually every few months :-/ It kind of sucks, but it's the only way I could figure out without leaving the keys to the castle on the server.

I suppose that for boringproxy it's a more acceptable choice, since an expired cert usually isn't a big deal, but still...

That being said, explicit adding of domains probably isn't a bad thing.
 
---

I actually gave up on boringproxy yesterday evening as I couldn't get the ACME to work at all; the server isn't exposed directly but via Varnish (since the server also runs other things), and I couldn't really figure out a good way to fix this, so I "just used ngrok" to test that one webhook (which was my goal in the first place before I got a bit distracted by this).

I kind of hate dealing with ACME; people say it's "easy", but in my experience it's not really; the whole verification stuff it pretty error-prone, especially in situations that deviate every-so-slightly from the intended setup, and a lot of the tooling/libraries aren't that great and don't give a whole lot of feedback on failures.

I'm not sure what a good solution is; I'm not sure if there is any, as a lot of this is inherent in how ACME works. But better feedback on *what exactly* it's doing and what exactly is going wrong would be a good start. I mean, that error message at the start is pretty long but also contains almost no information at all. I don't know if there's a way to tell certmagic to give more information

It's also difficult for me to test this stuff, since my ISP uses carrier grade NAT and I don't have a publicly accessible IP at all (it's shared with a bunch of other subscribers), so I need to work on it by sshing to a server :-/

**Comment by anderspitman** [comment]
> But people would need access to the server in order to do that, right?

Unfortunately no. It's not really clear from the demo, but boringproxy supports multi-tenancy. Admins can create as many users as they want, and each user can create tunnels for any domain, as long as ACME succeeds. I built it this way because I eventually want to offer boringproxy as a hosted service. The way you run GoatCounter was actually an inspiration to me. I like the idea of taking an open source project and charging people to run instances for them. I've actually been meaning to email you and ask you what lessons you've learned from this business model and whether you'd recommend it.

> The big downside of the dns-01 verification is that you need the modify the DNS records

Agreed. I'm planning to avoid the DNS challenge if at all possible.

> I actually gave up on boringproxy yesterday evening as I couldn't get the ACME to work at all; the server isn't exposed directly but via Varnish (since the server also runs other things), and I couldn't really figure out a good way to fix this, so I "just used ngrok" to test that one webhook (which was my goal in the first place before I got a bit distracted by this).

Sorry it didn't work out. It's definitely still beta quality. Running it behind another reverse proxy is totally untested. The intention is that for most setups, boringproxy is going to be your frontline proxy. But it's not ready for that yet either.

You're probably already aware of it since you found boringproxy, but there are more mature tunneling proxies on my list here: https://github.com/anderspitman/awesome-tunneling

> I kind of hate dealing with ACME; people say it's "easy", but in my experience it's not really

TBH my overall experience with certmagic at least has been very positive. I hate dependencies, but after trying to figure out what ACME would look like to implement myself took this one on gladly. And it really has just worked. It was a bit confusing to understand some of the semantics at first because it's relatively low level, but the design is solid.

> It's also difficult for me to test this stuff, since my ISP uses carrier grade NAT and I don't have a publicly accessible IP at all (it's shared with a bunch of other subscribers), so I need to work on it by sshing to a server :-/

I actually made boringproxy because I knew many people would need something like it to run the [federated storage protocol](https://github.com/gemdrive) I'm working on. People have big hard drives on their devices, but no way to access them easily behind NAT. I think tunneling could be a bridge to a true p2p internet. Once tunneling is robust and easy to do, developers can build p2p apps on top of it. Once the apps are there, hopefully there will be enough incentive to switch to ipv6 and drop NAT.

**Comment by anderspitman** [comment]
Oh and for a lot of ngrok-like tasks (webhooks being a great example), you should check out my https://patchbay.pub service. Even if you don't use it I think you'll like it :)

**Comment by arp242** [comment]
Ah right, I never got beyond trying to trying to start the server, so I never managed to see the web UI 😅

>      Sorry it didn't work out. It's definitely still beta quality. Running it behind another reverse proxy is totally untested. The intention is that for most setups, boringproxy is going to be your frontline proxy. But it's not ready for that yet either.

Well, it's not really your fault – it's just that ACME in general is pretty painful in these kind of scenarios. It took me quite a while to get goatcounter to work well with it too.

Anyway, I tried running it without proxy too (on a different port), but couldn't get that to work either; I'm not sure if the tls-alpn-01 challenge supports that or if it always tries to use port 80/443?

**Comment by anderspitman** [comment]
Yeah I'm pretty sure ACME (or maybe Let's Encrypt) specifies you must use ports 80/443.

**Comment by almereyda** [comment]
At times it may be sufficient to switch to the Let's Encrypt staging ACME endpoint for tunnels whose configuration is not yet stabilised, i.e. switching through different Client, Server or Passthrough conditions and hitting the rate limit, also keeping #42 in mind.

A simple DNS check prior requesting a certificate is also what Kubernetes' `cert-manager` does.

**Comment by anderspitman** [comment]
Yeah adding support for staging ACME would definitely be a good idea. It's very annoying when you have to wait a day to reuse a specific address. I'll add an issue.

**Comment by almereyda** [comment]
The bespoken issue can be found at #51, in case s/o wonders.
